src/brod_producers_sup.erl

```diff
@@ -0,0 +1,139 @@
+-define(DEFAULT_SUP2_RESTART_DELAY, 10).
+
+%% By default, restart partition producer process after a 2-seconds delay
+-define(DEFAULT_PRODUCER_RESTART_DELAY, 2).
+
+%%%_* APIs =====================================================================
+
+%% @doc Start a root producers supervisor,
+%%      per-topic supervisors and per-partition producer workers.
+%%      The config is passed down to the producers.
+%% @see brod_producer:start_link/4 for more details.
+%% @end
+-spec start_link(client_id(), [{topic(), producer_config()}]) -> {ok, pid()}.
+start_link(ClientId, Producers) ->
+  brod_supervisor:start_link({local, ?SUP}, ?MODULE,
+                             {?SUP, ClientId, Producers}).
+
+%% @doc Find a brod_producer process pid running under sup2
+%% @end
+-spec find_producer(pid(), topic(), partition()) ->
+                       {ok, pid()} | {error, any()}.
+find_producer(SupPid, Topic, Partition) ->
+  case brod_supervisor:find_child(SupPid, Topic) of
+    [] ->
+      %% no such topic worker started,
+      %% check sys.config or brod:start_link_client args
+      {error, {not_found, Topic}};
+    [Sup2Pid] ->
+      case is_alive(Sup2Pid) of
+        true ->
+          case brod_supervisor:find_child(Sup2Pid, Partition) of
+            [] ->
+              %% no such partition?
+              {error, {not_found, Topic, Partition}};
+            [Pid] ->
+              case is_alive(Pid) of
+                true  -> {ok, Pid};
+                false -> {error, restarting}
+              end
+          end;
+        false ->
+          {error, restarting}
+      end
+  end.
+
+%% @doc brod_supervisor callback.
+init({?SUP, ClientId, Producers}) ->
+  Children = [ producers_sup_spec(ClientId, TopicName, Config)
+             || {TopicName, Config} <- Producers ],
+  {ok, {{one_for_one, 0, 1}, Children}};
+init({?SUP2, _ClientId, _Topic, _Config}) ->
+  post_init.
+
+post_init({?SUP2, ClientId, Topic, Config}) ->
+  {ok, Partitions} = brod_client:get_partitions(ClientId, Topic),
+  Children = [ producer_spec(ClientId, Topic, Partition, Config)
+             || Partition <- Partitions ],
+  %% Producer may crash in case of exception in case of network failure,
+  %% or error code received in produce response (e.g. leader transition)
+  %% In any case, restart right away will erry likely fail again.
+  %% Hence set MaxR=0 here to cool-down for a configurable N-seconds
+  %% before supervisor tries to restart it.
+  {ok, {{one_for_one, 0, 1}, Children}};
+post_init(_) ->
+  ignore.
+
+producers_sup_spec(ClientId, TopicName, Config0) ->
+  DelaySecs = proplists:get_value(restart_delay_seconds, Config0,
+                                  ?DEFAULT_SUP2_RESTART_DELAY),
+  Config    = proplists:delete(restart_delay_seconds, Config0),
+  Args      = [?MODULE, {?SUP2, ClientId, TopicName, Config}],
+  { _Id       = TopicName
+  , _Start    = {brod_supervisor, start_link, Args}
+  , _Restart  = {permanent, DelaySecs}
+  , _Shutdown = infinity
```

Comment:
Recommended shutdown timeout for supervisors, they need to shut down children first.